File: 2026/09-September/22-Tuesday/solution.cpp
class Solution {
public:
    int numberOfWays(int n, int x) {
        const int MOD = 1e9 + 7;

        vector<long long> dp(n + 1, 0);
        dp[0] = 1;

        for (int i = 1; i <= n; i++) {
            long long power = 1;

            for (int j = 0; j < x; j++) {
                power *= i;
                if (power > n) break;
            }

            if (power > n)
                break;

            for (int sum = n; sum >= power; sum--) {
                dp[sum] = (dp[sum] + dp[sum - power]) % MOD;
            }
        }

        return dp[n];
    }
};

File: 2026/09-September/22-Tuesday/solution.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        MOD = 10**9 + 7

        dp = [0] * (n + 1)
        dp[0] = 1

        for i in range(1, n + 1):
            power = i ** x

            if power > n:
                break

            for s in range(n, power - 1, -1):
                dp[s] = (dp[s] + dp[s - power]) % MOD

        return dp[n]
